system/Files/FileCollection.php

```diff
@@ -0,0 +1,396 @@
+<?php
+
+/**
+ * This file is part of the CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\Files;
+
+use CodeIgniter\Files\Exceptions\FileException;
+use CodeIgniter\Files\Exceptions\FileNotFoundException;
+use Countable;
+use Generator;
+use InvalidArgumentException;
+use IteratorAggregate;
+use Traversable;
+
+/**
+ * File Collection Class
+ *
+ * Representation for a group of files, with utilities for locating,
+ * filtering, and ordering them.
+ */
+class FileCollection implements Countable, IteratorAggregate
+{
+	/**
+	 * The current list of file paths.
+	 *
+	 * @var string[]
+	 */
+	protected $files = [];
+
+	//--------------------------------------------------------------------
+
+	/**
+	 * Resolves a full path and verifies it is an actual directory.
+	 *
+	 * @param string $directory
+	 *
+	 * @return string
+	 *
+	 * @throws FileException
+	 */
+	final protected static function resolveDirectory(string $directory): string
```

Comment:
`final` method doesn't need `protected`:

```suggestion
	final static function resolveDirectory(string $directory): string
```

Reply:
Why not just change to `private` so it can't be accessed by child classes and no need for `final`?

Reply:
@paulbalandan I just realized that 👍

Reply:
`Publisher` needs this method (which is where it came from) so when it moved to `FileCollection` I changed it from private to protected. We could make it private and give `Publisher` its own version, but I think it likely that if one child needs it then others might. 

An alternate approach would be to make these support methods into functions and add them to Filesystem Helper. I generally shy away from helper bloat but at least some of these seem generic enough.